Sources/NIOCore/ByteBuffer-QUICLengthPrefix.swift

```diff
@@ -0,0 +1,238 @@
+//
+// Copyright (c) 2024 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if os(Windows)
+import ucrt
+#elseif canImport(Darwin)
+import Darwin
+#elseif canImport(Glibc)
+import Glibc
+#elseif canImport(Musl)
+import Musl
+#elseif canImport(Bionic)
+import Bionic
+#else
+#error("The Byte Buffer module was unable to identify your C library.")
+#endif
+
+extension ByteBuffer {
+    /// Read a QUIC variable-length integer, moving the `readerIndex` appropriately
+    /// If there are not enough bytes, nil is returned
+    ///
+    /// [RFC 9000 § 16](https://www.rfc-editor.org/rfc/rfc9000.html#section-16)
+    public mutating func readQUICVariableLengthInteger() -> Int? {
+        guard let firstByte = self.getInteger(at: self.readerIndex, as: UInt8.self) else {
+            return nil
+        }
+
+        // Look at the first two bits to work out the length, then read that, mask off the top two bits, and
+        // extend to integer.
+        switch firstByte & 0xC0 {
+        case 0x00:
+            // Easy case.
+            self.moveReaderIndex(forwardBy: 1)
+            return Int(firstByte & ~0xC0)
+        case 0x40:
+            // Length is two bytes long, read the next one.
+            return self.readInteger(as: UInt16.self).map { Int($0 & ~(0xC0 << 8)) }
+        case 0x80:
+            // Length is 4 bytes long.
+            return self.readInteger(as: UInt32.self).map { Int($0 & ~(0xC0 << 24)) }
+        case 0xC0:
+            // Length is 8 bytes long.
+            return self.readInteger(as: UInt64.self).map { Int($0 & ~(0xC0 << 56)) }
+        default:
+            fatalError("Unreachable")
+        }
+    }
+
+    /// Write a QUIC variable-length integer.
+    ///
+    /// [RFC 9000 § 16](https://www.rfc-editor.org/rfc/rfc9000.html#section-16)
+    ///
+    /// - Returns: The number of bytes written
+    @discardableResult
+    public mutating func writeQUICVariableLengthInteger<Integer: FixedWidthInteger>(_ value: Integer) -> Int {
+        switch value {
+        case 0..<63:
+            // Easy, store the value. The top two bits are 0 so we don't need to do any masking.
+            return self.writeInteger(UInt8(truncatingIfNeeded: value))
+        case 0..<16383:
+            // Set the top two bit mask, then write the value.
+            let value = UInt16(truncatingIfNeeded: value) | (0x40 << 8)
+            return self.writeInteger(value)
+        case 0..<1_073_741_823:
+            // Set the top two bit mask, then write the value.
+            let value = UInt32(truncatingIfNeeded: value) | (0x80 << 24)
+            return self.writeInteger(value)
+        case 0..<4_611_686_018_427_387_903:
```

Comment:
I don't think this works with smaller integer sizes. Can you write a test that tries to call this with a `UInt8`, validates that it doesn't work, and then fix it?